<a href="https://colab.research.google.com/github/Thitichaya2000/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. วันธรรมดาหรือวันเสาร์-อาทิตย์มีเครื่องบินขาเข้ามากกว่ากัน (วันจันทร์ = 1, วันอังคาร = 2, ...)
6. เรียงลำดับ ชื่อสนามบิน ที่มีรหัสสนามบิน 5 อันดับแรก

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [4]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [5]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [10]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [11]:
gen_exams('623020520-0')

รหัสนักศึกษา 623020520-0 ทำข้อ 0 และ 5


##0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด

In [25]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินปลายทาง',right_on='รหัสสนามบิน') # how = ตารางหลัก left_on ตารางซ้ายของตารางหลัก ผูกกับ right_on id
merged_table

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 11 columns]

In [50]:
merged_table.dropna() ##เช็ค missing

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 11 columns]

In [51]:
merged_table.isnull() .any() #เช็คดู ถ้ามี true แสดงว่ามีค่า missing

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินต้นทาง     False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
รหัสสนามบิน           False
เมือง                 False
รัฐ                   False
ชื่อสนามบิน           False
dtype: bool

In [26]:
merged_table.groupby('รัฐ').count() #ตรวจดูจำนวนเที่ยวบินของแต่ละรัฐ

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
AK         7711          7711       7711  ...         7711    7711         7711
AZ       100824        100824     100824  ...       100824  100824       100824
CA       380429        380429     380429  ...       380429  380429       380429
CO        96919         96919      96919  ...        96919   96919        96919
CT        13071         13071      13071  ...        13071   13071        13071
DC        80666         80666      80666  ...        80666   80666        80666
FL       208291        208291     208291  ...       208291  208291       208291
GA       148524        148524     148524  ...       148524  148524       148524
HI        28402         28402      28402  ...        28402   28402        28402
IL       176857        176857     176857  ...       176857  176857       176857
IN        18085         18085      18085  ...        18085   18085        18085
KY        10945         10945      10945  ...        10945   10945        10945
LA        25189         25189      25189  ...        25189   25189        25189
MA        66833         66833      66833  ...        66833   66833        66833
MD        51679         51679      51679  ...        51679   51679        51679
MI        62728         62728      62728  ...        62728   62728        62728
MN        60015         60015      60015  ...        60015   60015        60015
MO        63593         63593      63593  ...        63593   63593        63593
NC       104839        104839     104839  ...       104839  104839       104839
NE        13516         13516      13516  ...        13516   13516        13516
NJ        64071         64071      64071  ...        64071   64071        64071
NM        17558         17558      17558  ...        17558   17558        17558
NV        88179         88179      88179  ...        88179   88179        88179
NY       132522        132522     132522  ...       132522  132522       132522
OH        66236         66236      66236  ...        66236   66236        66236
OK        13902         13902      13902  ...        13902   13902        13902
OR        30579         30579      30579  ...        30579   30579        30579
PA        67955         67955      67955  ...        67955   67955        67955
PR        13787         13787      13787  ...        13787   13787        13787
RI         9324          9324       9324  ...         9324    9324         9324
TN        53470         53470      53470  ...        53470   53470        53470
TX       275328        275328     275328  ...       275328  275328       275328
UT        47410         47410      47410  ...        47410   47410        47410
VA        21539         21539      21539  ...        21539   21539        21539
WA        60237         60237      60237  ...        60237   60237        60237
WI        21005         21005      21005  ...        21005   21005        21005

[36 rows x 10 columns]

In [60]:
max = merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) #รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุด
max.iloc[:1,4:5]

รหัสสนามบินปลายทาง
รัฐ                    
CA               380429

In [61]:
min = merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False)  #รัฐที่มีจำนวนเครื่องบินขาเข้าน้อยที่สุด
min.iloc[-1:,4:5]

รหัสสนามบินปลายทาง
รัฐ                    
AK                 7711

##5.วันธรรมดาหรือวันเสาร์-อาทิตย์มีเครื่องบินขาเข้ามากกว่ากัน (วันจันทร์ = 1, วันอังคาร = 2, ...)


In [71]:
merged_table.groupby('วันในสัปดาห์').count()

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
1                 407837     407837  ...  407837       407837
2                 397594     397594  ...  397594       397594
3                 403072     403072  ...  403072       403072
4                 406563     406563  ...  406563       406563
5                 396387     396387  ...  396387       396387
6                 318537     318537  ...  318537       318537
7                 372228     372228  ...  372228       372228

[7 rows x 10 columns]

In [73]:
week_Arrival_plane = merged_table.groupby('วันในสัปดาห์').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) #ดูจำนวนเครื่องบินขาเข้าในแต่ละวัน
week_Arrival_plane

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
1                 407837     407837  ...  407837       407837
4                 406563     406563  ...  406563       406563
3                 403072     403072  ...  403072       403072
2                 397594     397594  ...  397594       397594
5                 396387     396387  ...  396387       396387
7                 372228     372228  ...  372228       372228
6                 318537     318537  ...  318537       318537

[7 rows x 10 columns]

In [74]:
Arrival_on_weekdays = week_Arrival_plane.iloc[:5,3:4].sum()  #หาจำนวนเครื่องบินขาเข้าในวันธรรมดา
Arrival_on_weekdays

รหัสสนามบินปลายทาง    2011453
dtype: int64

In [76]:
Holiday_arrivals = week_Arrival_plane.iloc[5:7,3:4].sum() #หาจำนวนเครื่องบินขาเข้าในวัน เสาร์-อาทิตย์
Holiday_arrivals

รหัสสนามบินปลายทาง    690765
dtype: int64

วันธรรมดามีเครื่องบินขาเข้า  2011453

วันเสาร์-อาทิตย์มีเครื่องบินขาเข้า 690765

**ดังนั้น วันธรรมดามีเครื่องบินขาเข้ามากกว่าวันเสาร์-อาทิตย์**

##6.เรียงลำดับ ชื่อสนามบิน ที่มีรหัสสนามบิน 5 อันดับแรก

In [45]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินปลายทาง',right_on='รหัสสนามบิน') # how = ตารางหลัก left_on ตารางซ้ายของตารางหลัก ผูกกับ right_on id
merged_table

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 11 columns]

In [67]:
merged_table.groupby('ชื่อสนามบิน').count() 

วันในเดือน  ...    รัฐ
ชื่อสนามบิน                                                     ...       
Albuquerque International Sunport                        17558  ...  17558
Austin - Bergstrom International                         27009  ...  27009
Baltimore/Washington International Thurgood Mar...       51679  ...  51679
Bob Hope                                                 14187  ...  14187
Bradley International                                    13071  ...  13071
...                                                        ...  ...    ...
Theodore Francis Green State                              9324  ...   9324
Tucson International                                     11010  ...  11010
Washington Dulles International                          37629  ...  37629
Will Rogers World                                        13902  ...  13902
William P Hobby                                          28905  ...  28905

[70 rows x 10 columns]

In [68]:
airport_name =merged_table.groupby('ชื่อสนามบิน').count().sort_values(by=['รหัสสนามบิน'],ascending=False)
airport_name 

วันในเดือน  ...     รัฐ
ชื่อสนามบิน                                           ...        
Hartsfield-Jackson Atlanta International      148524  ...  148524
Chicago O'Hare International                  127341  ...  127341
Los Angeles International                     118274  ...  118274
Dallas/Fort Worth International               103939  ...  103939
Denver International                           96919  ...   96919
...                                              ...  ...     ...
Palm Beach International                       10103  ...   10103
Norfolk International                          10035  ...   10035
Theodore Francis Green State                    9324  ...    9324
Long Beach Airport                              8171  ...    8171
Ted Stevens Anchorage International             7711  ...    7711

[70 rows x 10 columns]

In [70]:
#เรียงลำดับ ชื่อสนามบิน ที่มีรหัสสนามบิน 5 อันดับแรก
airport_name_cut  = airport_name .iloc[:5,7:8] 
airport_name_cut

รหัสสนามบิน
ชื่อสนามบิน                                          
Hartsfield-Jackson Atlanta International       148524
Chicago O'Hare International                   127341
Los Angeles International                      118274
Dallas/Fort Worth International                103939
Denver International                            96919